apps/indexer/lib/indexer/fetcher/on_demand/token_balance.ex

```diff
@@ -39,39 +44,80 @@ defmodule Indexer.Fetcher.OnDemand.TokenBalance do
           String.t(),
           Decimal.t() | nil,
           non_neg_integer()
-        ) :: {:ok, pid}
+        ) :: :ok
   def trigger_historic_fetch(address_hash, contract_address_hash, token_type, token_id, block_number) do
-    Task.start(fn ->
-      do_trigger_historic_fetch(address_hash, contract_address_hash, token_type, token_id, block_number)
-    end)
+    unless __MODULE__.Supervisor.disabled?() do
+      BufferedTask.buffer(
+        __MODULE__,
+        [{:historic_fetch, {address_hash, contract_address_hash, token_type, token_id, block_number}}],
+        false
+      )
+    end
   end
 
-  ## Implementation
+  @doc false
+  def child_spec([init_options, gen_server_options]) do
+    merged_init_opts =
+      @defaults
+      |> Keyword.merge(init_options)
+      |> Keyword.put(:state, %{})
 
-  defp do_trigger_fetch(address_hash, latest_block_number, stale_balance_window)
-       when not is_nil(address_hash) do
-    stale_current_token_balances =
-      address_hash
-      |> Chain.fetch_last_token_balances_include_unfetched()
-      |> delete_invalid_balances()
-      |> Enum.filter(fn current_token_balance -> current_token_balance.block_number < stale_balance_window end)
+    Supervisor.child_spec({BufferedTask, [{__MODULE__, merged_init_opts}, gen_server_options]}, id: __MODULE__)
+  end
 
-    if Enum.empty?(stale_current_token_balances) do
-      :current
-    else
-      fetch_and_update(latest_block_number, address_hash, stale_current_token_balances)
-    end
+  @impl BufferedTask
+  def init(initial, _, _) do
+    initial
+  end
+
+  @impl BufferedTask
+  def run(entries, _) do
+    {fetch_params, historic_fetch_params} =
+      Enum.reduce(entries, {[], []}, fn
+        {:fetch, address_hash}, {fetch_acc, historic_acc} ->
+          {[address_hash | fetch_acc], historic_acc}
+
+        {:historic_fetch, params}, {fetch_acc, historic_acc} ->
+          {fetch_acc, [params | historic_acc]}
+      end)
```

Comment:
for consistency here we could apply the similar approach:
```suggestion
   entries_by_type = Enum.group_by(entries, &elem(&1, 0), &elem(&1, 1))
    {fetch_params, historic_fetch_params} = {entries_by_type[:fetch], entries_by_type[:historic_fetch]}
```

Reply:
Processed in 512f9f19ac9b925352d5729128ae683c124a29d0